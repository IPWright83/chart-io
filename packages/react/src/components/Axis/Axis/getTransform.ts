import { Margin, Position } from "../../../types";

/**
 * Obtain the transform for the Axis group
 * @param  position     The position of the axis [left, right, top, bottom]
 * @param  width        The width for the chart
 * @param  height       The height for the chart
 * @param  margin       The margin object for the chart
 * @return              The transform
 */
const getTransform = (position: Position, width: number, height: number, margin: Margin): string => {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    switch (position) {
        case "left":
            return `translate(${margin.left}, 0)`;
        case "right":
            return `translate(${width - margin.right}, 0)`;
        case "top":
            return `translate(0, ${margin.top})`;
        case "bottom":
            return `translate(0, ${height - margin.bottom})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
};

export { getTransform };
