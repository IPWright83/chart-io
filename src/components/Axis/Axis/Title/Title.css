.axis-title {
    text-anchor: middle;
    font-family: sans-serif;
    font-size: 10px;
    user-select: none;
}
