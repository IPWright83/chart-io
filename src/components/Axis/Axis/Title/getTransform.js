/**
 * Obtain the transform for the Axis title
 * @param  {String} position     The position of the axis [left, right, top, bottom]
 * @param  {Number} width        The width for the chart
 * @param  {Number} height       The height for the chart
 * @param  {Number} margin       The margin object for the chart
 * @return {String}              The transform
 */
const getTransform = (position, width, height, margin) => {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    const actualWidth = width - margin.left - margin.right;
    const actualHeight = height - margin.top - margin.bottom;
    const halfWidth = actualWidth / 2;
    const halfHeight = actualHeight / 2;

    switch (position) {
        case "left":
            return `translate(${margin.left / 2 - 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "right":
            return `translate(${width - margin.right / 2 + 25}, ${halfHeight + margin.top / 2})rotate(270)`;
        case "top":
            return `translate(${halfWidth + margin.left / 2}, ${margin.top - 25})`;
        case "bottom":
            return `translate(${halfWidth + margin.left / 2}, ${height - margin.bottom + 35})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
};

export { getTransform };
