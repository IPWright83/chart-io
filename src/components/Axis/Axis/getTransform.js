/**
 * Obtain the transform for the Axis group
 * @param  {String} position     The position of the axis [left, right, top, bottom]
 * @param  {Number} width        The width for the chart
 * @param  {Number} height       The height for the chart
 * @param  {Number} margin       The margin object for the chart
 * @return {String}              The transform
 */
const getTransform = (position, width, height, margin) => {
    if (width === 0 || height === 0) {
        return "translate(0, 0)";
    }

    switch (position) {
        case "left":
            return `translate(${margin.left}, 0)`;
        case "right":
            return `translate(${width - margin.right}, 0)`;
        case "top":
            return `translate(0, ${margin.top})`;
        case "bottom":
            return `translate(0, ${height - margin.bottom})`;
        default:
            throw new Error(`Invalid position: ${position}`);
    }
};

export { getTransform };
