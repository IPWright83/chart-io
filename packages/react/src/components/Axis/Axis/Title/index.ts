import { Title } from "./Title";

import { getTransform } from "./getTransform";

// @ts-ignore Ignore additional exported helpers
Title.getTransform = getTransform;

export { Title };
