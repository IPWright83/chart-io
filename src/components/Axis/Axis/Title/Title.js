import "./Title.css";

import React from "react";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";

import { getTransform } from "./getTransform";

import { chartSelectors } from "../../../../store";

const Title = ({ position, title, fields }) => {
    const width = useSelector((s) => chartSelectors.dimensions.width(s));
    const height = useSelector((s) => chartSelectors.dimensions.height(s));
    const margin = useSelector((s) => chartSelectors.dimensions.margin(s));

    const transform = getTransform(position, width, height, margin);
    const text = title || fields.filter((f) => !!f).join(",");

    return (
        <text className="axis-title" transform={transform}>
            {text}
        </text>
    );
};

Title.propTypes = {
    /**
     * The position of the axis [top, bottom, left, right]
     * @type {String}
     */
    position: PropTypes.oneOf(["top", "bottom", "left", "right"]),
    /**
     * The keys of the fields that will share this scale
     * @type {String[]}
     */
    fields: PropTypes.arrayOf(PropTypes.string).isRequired,
    /**
     * A title for the Axis
     * @type {String}
     */
    title: PropTypes.string,
};

export { Title };
