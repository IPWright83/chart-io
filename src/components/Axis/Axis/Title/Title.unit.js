import React from "react";
import { Provider } from "react-redux";

import TestRenderer from "react-test-renderer";

import { Title } from "./Title";
import { getTransform } from "./getTransform";

describe("Title", () => {
    const width = 1000;
    const height = 500;
    const margin = { left: 50, right: 60, top: 30, bottom: 40 };

    const store = {
        getState: () => ({
            chart: {
                dimensions: {
                    width,
                    height,
                    margin,
                },
            },
        }),
        dispatch: () => {},
        subscribe: () => {},
    };

    describe("getTransform", () => {
        describe("should return correct transform", () => {
            it("for left axis", () => {
                expect(getTransform("left", width, height, margin)).toBe("translate(0, 230)rotate(270)");
            });
            it("for right axis", () => {
                expect(getTransform("right", width, height, margin)).toBe("translate(995, 230)rotate(270)");
            });
        });

        describe("should return correct transform", () => {
            it("for top axis", () => {
                expect(getTransform("top", width, height, margin)).toBe("translate(470, 5)");
            });
            it("for bottom axis", () => {
                expect(getTransform("bottom", width, height, margin)).toBe("translate(470, 495)");
            });
        });

        it("throws an error with an invalid position", () => {
            expect(() => getTransform("invalid", width, height, margin)).toThrow();
        });
    });

    describe("component", () => {
        describe("with title set", () => {
            it("renders horizontal title", async () => {
                let root;

                TestRenderer.act(() => {
                    root = TestRenderer.create(
                        <Provider store={store}>
                            <Title position="bottom" title="horizontal" fields={["a", "b"]} />
                        </Provider>
                    ).toJSON();
                });

                expect(root).toMatchSnapshot();
            });

            it("renders vertical title", async () => {
                let root;

                TestRenderer.act(() => {
                    root = TestRenderer.create(
                        <Provider store={store}>
                            <Title position="left" title="vertical" fields={["a", "b"]} />
                        </Provider>
                    ).toJSON();
                });

                expect(root).toMatchSnapshot();
            });
        });

        describe("with no explicit title", () => {
            it("renders fields as horizontal title", async () => {
                let root;

                TestRenderer.act(() => {
                    root = TestRenderer.create(
                        <Provider store={store}>
                            <Title position="bottom" fields={["a", "b"]} />
                        </Provider>
                    ).toJSON();
                });

                expect(root).toMatchSnapshot();
            });

            it("renders fields as vertical title", async () => {
                let root;

                TestRenderer.act(() => {
                    root = TestRenderer.create(
                        <Provider store={store}>
                            <Title position="left" fields={["a", "b"]} />
                        </Provider>
                    ).toJSON();
                });

                expect(root).toMatchSnapshot();
            });
        });
    });
});
