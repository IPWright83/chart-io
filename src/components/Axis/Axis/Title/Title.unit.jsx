import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";

import { createMockStore } from "../../../../testUtils";

import { Title } from "./Title";
import { getTransform } from "./getTransform";

describe("Title", () => {
    const width = 1000;
    const height = 500;
    const margin = { left: 50, right: 60, top: 30, bottom: 40 };

    const store = createMockStore({
        chart: {
            dimensions: {
                width,
                height,
                margin,
            },
        },
    });

    describe("getTransform", () => {
        describe("should return correct transform", () => {
            it("for left axis", () => {
                expect(getTransform("left", width, height, margin)).toBe("translate(0, 230)rotate(270)");
            });
            it("for right axis", () => {
                expect(getTransform("right", width, height, margin)).toBe("translate(995, 230)rotate(270)");
            });
        });

        describe("should return correct transform", () => {
            it("for top axis", () => {
                expect(getTransform("top", width, height, margin)).toBe("translate(470, 5)");
            });
            it("for bottom axis", () => {
                expect(getTransform("bottom", width, height, margin)).toBe("translate(470, 495)");
            });
        });

        describe("should return no translation", () => {
            it("for a 0 width", () => {
                expect(getTransform("top", 0, height, margin)).toBe("translate(0, 0)");
            });
            it("for a 0 height", () => {
                expect(getTransform("top", width, 0, margin)).toBe("translate(0, 0)");
            });
        });

        it("throws an error with an invalid position", () => {
            expect(() => getTransform("invalid", width, height, margin)).toThrow();
        });
    });

    describe("component", () => {
        describe("with title set", () => {
            it("renders horizontal title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="bottom" title="horizontal" fields={["a", "b"]} />
                        </svg>
                    </Provider>,
                );

                expect(asFragment()).toMatchSnapshot();
            });

            it("renders vertical title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="left" title="vertical" fields={["a", "b"]} />
                        </svg>
                    </Provider>,
                );

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("with no explicit title", () => {
            it("renders fields as horizontal title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="bottom" fields={["a", "b"]} />
                        </svg>
                    </Provider>,
                );

                expect(asFragment()).toMatchSnapshot();
            });

            it("renders fields as vertical title", async () => {
                const { asFragment } = render(
                    <Provider store={store}>
                        <svg>
                            <Title position="left" fields={["a", "b"]} />
                        </svg>
                    </Provider>,
                );

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
