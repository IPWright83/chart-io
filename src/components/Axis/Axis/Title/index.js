import { Title } from "./Title";

import { getTransform } from "./getTransform";
Title.getTransform = getTransform;

export { Title };
