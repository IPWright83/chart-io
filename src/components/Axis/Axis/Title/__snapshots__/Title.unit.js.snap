// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component with no explicit title renders fields as horizontal title 1`] = `null`;

exports[`Title component with no explicit title renders fields as vertical title 1`] = `null`;

exports[`Title component with title set renders horizontal title 1`] = `null`;

exports[`Title component with title set renders vertical title 1`] = `null`;
